### ISSUE TITLE
Wait for services to be ready & available (docker)

### ISSUE THREAD

**Issue by Flink** [ayufan/gitlab-ci-multi-runner#9]
Just found out another little bug :wink: 

I’m using bundler and cache the gems in `/cache`. I’m also using PG as my DB for the tests (but the problem will be the same with MySQL or Mongo I think). When I retry the build (failed or not), the bundle phase is so fast that the next instruction is executed right away (`rake db:setup`) and I suspect strongly that PG in its container is not completely started leading to a connect error.

After some (quick) research, it seems docker doesn’t know how to handle that kind of thing. But you could check that the services container are started before starting the main container (or executing the build in it).

I’ve found some links that suggest various solutions:
- https://github.com/docker/compose/issues/374#issuecomment-67790672
- https://github.com/aanand/docker-wait
- https://stackoverflow.com/questions/21183088/how-can-i-wait-for-a-docker-container-to-be-up-and-running

**Comment by ayufan** [comment]
I guess it's unresolvable. There are two options: 1. add some sleep before, 2. inject `https://github.com/aanand/docker-wait/blob/master/wait` into build script, but grep over all exposed services.

Maybe you would have some time to add one of this into runner's code?

**Comment by Flink** [comment]
I can try but I never programmed in Go :)

**Comment by ayufan** [comment]
It's really simple. I did this project as learning opportunity :) Before this one I made one really small app.

**Comment by Flink** [comment]
Ok then I’ll give it a try ;)

**Comment by ayufan** [comment]
Some maybe configurable sleep would be enough?

**Comment by Flink** [comment]
It could but it won’t always work. Checking for the link to be up is much more reliable I think.

**Comment by ayufan** [comment]
So maybe rework this script (https://github.com/aanand/docker-wait/blob/master/wait) to inject it into build? Somewhere around there: https://github.com/ayufan/gitlab-ci-multi-runner/blob/master/common/build.go#L138

**Comment by ayufan** [comment]
If you want to work on dev sources simply do: 
```
go get github.com/ayufan/gitlab-ci-multi-runner
cd $GOPATH/src/github.com/ayufan/gitlab-ci-multi-runner
go run main.go run
...
go install
```

**Comment by Flink** [comment]
Ok thank you!
And yes I was thinking about using this docker-wait utility :)

**Comment by Flink** [comment]
Hmm I think we could inject docker-wait just after each service run. Somewhere in https://github.com/ayufan/gitlab-ci-multi-runner/blob/master/executors/docker/executor_docker.go ? But for now I don’t see really where the `docker run` command for services is invoked.

**Comment by ayufan** [comment]
Runner talks directly to the docker daemon using Client API.

**Comment by Flink** [comment]
Ok, I’m hacking something, I’ll tell you if it works ;)

**Comment by Flink** [comment]
I may have something but I can’t test my code… When I launch `go run main.go run`, well it executes the code in the repository but not on the disk, I don’t understand how it works :(

**Comment by Flink** [comment]
Ok finally I understood how it works ;) Much complicated than ruby :)
So for now I have something that doesn’t work at all but the image is pulled, container is created but it seems it does not block the execution of the rest of the code… So pretty useless. And those containers aren’t destroyed when the build is finished.
If you want to see what I have for now: https://github.com/Flink/gitlab-ci-multi-runner/commit/6cedb9aef7e2e9edc17f667aee56252faae833cc

**Comment by ayufan** [comment]
It should work, but your are still missing wait for exit and later removal of container. Wouldn't it be simpler to inject just bash script into executed script? :)

**Comment by Flink** [comment]
Yes it should but we'll have to rely on third party tools which may be or may be not installed in the base image. For example the ruby:2.2 image has curl but not nc. But we could test which program is available and use it I suppose :)

**Comment by Flink** [comment]
Ok it’s done :)
I’ve made a PR.

**Comment by ayufan** [comment]
I renamed it to WaitForServicesTimeout which is by default set to 30s. If it timeouts it just prints warning message and build continues. It should get released as 0.1.10 within a few minutes
.

**Comment by Flink** [comment]
Ok that’s great, thank you :)
